
---
title: "scDIV: Single Cell RNA Sequencing Data Demultiplexing using Interindividual Variations"
author: "Isar Nassiri"
date: "January 23, 2023"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{scDIV: Single Cell RNA Sequencing Data Demultiplexing using Interindividual Variations\}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

* [Introduction](#introduction)
* [Installation](#installation)
* [Step 1: Infer genetic variants from scRNA-seq data ](#Step1)
* [Step 2: Demultiplex pooled samples ](#Step2)
* [Step 3: Generate gene-cell count matrix ](#Step3)
* [Step 4: Single Cell Data Quality Control ](#Step4)
* [Step 5: Gene Expression Recovery ](#Step5)
* [Step 6: Inter-individual Differential gene Correlation Analysis (IDCA) ](#Step6)
* [Step 7: Visualization of IDCA outputs ](#Step7)
* [Step 8: Expression Aware Demultiplexing per Donor Pair ](#Step8)
* [Step 9: Expression Aware Demultiplexing per sample pool ](#Step9)
* [Citation](#Citation)

<a name="introduction"/>

### Introduction

scQCEA (acronym of the single-cell RNA sequencing Quality Control and Enrichment Analysis) is an R package for annotation and quality control report of scRNA-Seq profiles, which performs a probabilistic assignment of the reference cell types to identify clusters, before downstream analysis such as gene network inference. scQCEA provides automated cell type annotation on scRNA-seq data and identifies differential patterns in gene expression. scQCEA generates an interactive report of quality control metrics which allows visual evaluation of QC metrics, objective selection of insightful optimal cluster numbers and discrimination between true variation and background noise. 

Please see the [`manual`](https://isarnassiri.github.io/scQCEA/) for the usage of scQCEA including the explanations of the HTML report and how to prepare data input files.

<a name="installation"/>

### Installation
1. Install the R [(LINK)](https://cran.r-project.org/)
2. Install the free version of rStudio [(LINK)](https://www.rstudio.com/products/rstudio/download/)
3. Run the following command in rStudio to install scQCEA as an R package:

```{r,eval=FALSE}
install_github("isarnassiri/scQCEA")
```

<a name="Step1"/>

### Step 1: Infer genetic variants from scRNA-seq data 
cellsnp-lite is used to pileup the expressed alleles in single-cell data, which can be directly used for donor deconvolution in multiplexed single-cell RNA-seq data, which assigns cells to donors without genotyping reference [(LINK)](https://github.com/single-cell-genetics/cellsnp-lite). 

cellsnp-lite gets bam file and list of barcodes as variable inputs, a Variant Call Format (vcf) file listing all candidate SNPs (regionsVCF) as backend input variable:
  
```{r,eval=FALSE}
cellsnp-lite -s possorted_genome_bam.bam -b barcodes.tsv.gz -O FOLDER-NAME -R regionsVCF -p 22 --minMAF 0.05 --minCOUNT 10 --gzip 
```

cellsnp-lite generates a vcf file including called genetic variants as follows:
  
| ![Figure 1](cellsnp-lite.png){width=50%} | 
|:--:| 
| *Figure 1. Example of vcf file generated by cellsnp-lite.* |

<a name="Step2"/>

### Step 2: Demultiplex pooled samples 
We use Vireo (Variational Inference for Reconstructing Ensemble Origin) for donor deconvolution using expressed SNPs in multiplexed scRNA-seq data [(LINK)](https://vireosnp.readthedocs.io/en/latest/).

Vireo gets s variants info file provided by cellsnp-lite as an input: 
  
```{r,eval=FALSE}
vireo -c input-vcf-file -o output-folder --randSeed 2 -N Number-of-donors -t GP  
```

We use "donor_ids.tsv" file from outputs of Vireo for downstream analysis:
  
| ![Figure 2](vireo.png){width=50%} | 
|:--:| 
| *Figure 2. Example of output from Vireo.* |

<a name="Step3"/>

### Step 3: Generate gene-cell count matrix
Raw count data from 10X CellRanger (outs/read_count.csv) or other single-cell experiments has the gene as a row (the gene name should be the human or mouse Ensembl gene ID) and the cell as a column. You can convert an HDF5 Feature-Barcode Matrix [(LINK)](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices) to a gene-cell count matrix using the cellranger mat2csv [(LINK)](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices#mat2csv) command provided by 10Xgenomics. The cells in the `read_count.csv` file are from the filtered feature-barcode matrix generated by the cell ranger.

A filtered feature-barcode matrix generated by the cell ranger can be converted from HDF5 feature-barcode matrix to a gene-cell count matrix (read_count.csv) using the cellranger mat2csv (command provided by 10Xgenomics) as follows:

```{r,eval=FALSE}
cellranger mat2csv filtered_feature_bc_matrix.h5 read_count.csv
```

<a name="Step4"/>

### Step 4: Single Cell Data Quality Control
In this step, we start to use a function from scDIV package. We use `scQC()` function to identify potential doublet cells based on the local density of simulated doublet expression profiles as follows:

```{r,eval=FALSE}
library("scDIV")
csQCEAdir <- system.file("extdata", package = "scDIV")
scQC( InputDir = csQCEAdir ) 
```
You can find the results in the QC/ folder under the title of 'Cells_passed_QC_noDoublet.txt'.

<a name="Step5"/>

### Step 5: Gene Expression Recovery
The `GeneExpressionRecovery()` function uses SAVER (single-cell analysis via expression recovery), an expression recovery method for unique molecule index (UMI)-based scRNA-seq data to provide accurate expression estimates for all genes in a scRNA-seq profile.

```{r,eval=FALSE}
library("scDIV")
csQCEAdir <- system.file("extdata", package = "scDIV")
Donors='donor6_donor2'
FC='FAI5649A17'
GeneExpressionRecovery( InputDir = csQCEAdir, Donors = Donors, FC = FC )
```

You can find the results in the SAVER/ folder with 'AssignedCells.txt' and 'AllCells.txt' extensions.

<a name="Step6"/>

### Step 6: Inter-individual Differential gene Correlation Analysis (IDCA)
The `IDCA()` function uses correlation coefficients and performed Inter-individual Differential gene Correlation Analysis (IDCA) for two donors (D1 and D2) and genes (G1 and G2).

```{r,eval=FALSE}
library("scDIV")
ERP = "donor6_donor2_FAI5649A17_AssignedCells.txt"
Donors='donor6_donor2'
FC='FAI5649A17'
InputDir = system.file("extdata", package = "scDIV")
IDCA( InputDir, Donors, FC, ERP, TEST = T )
```

You can find the results in the IDCA_Analysis/ folder.

<a name="Step7"/>

### Step 7: Visualization of IDCA outputs
The `IDCAvis()` function visualizes the outputs of IDC analysis.

```{r,eval=FALSE}
library("scDIV")
InputDir = system.file("extdata", package = "scDIV")
IDCAvis( InputDir )
```

You can find the results in the IDCA_Analysis/IDCA_Plots/ as pdf file(s).

<a name="Step8"/>

### Step 8: Expression Aware Demultiplexing per Donor Pair
The `EADDonorPair()` function uses inter-individual differential co-expression patterns for demultiplexing per donor pair.

```{r,eval=FALSE}
library("scDIV")
InputDir = system.file("extdata", package = "scDIV")
EADDonorPair( InputDir )
```

You can find the results in the IDCA_Analysis/Expression_Aware_Cell_Assignment/ folder called "Expression_Aware_Cell_Assignment.txt".

<a name="Step9"/>

### Step 9: Expression Aware Demultiplexing per sample pool
The `EADDonorPair()` function uses inter-individual differential co-expression patterns for demultiplexing the pooled samples

```{r,eval=FALSE}
library("scDIV")
InputDir = system.file("extdata", package = "scDIV")
EADDonorPair( InputDir )
```

You can find the results in the IDCA_Analysis/Expression_Aware_Cell_Assignment/ folder called "Results_Expression_Aware_Cell_Assignment.txt" and "Summary.txt".

<a name="Citation"/>

### Citation
Isar Nassiri, Benjamin Fairfax, Andrew J Kwok, Aneesha Bhandari, Katherine Bull, Angela Lee, Yanxia Wu, Julian Knight, David Buck, Paolo Piazza. Demultiplexing of Single Cell RNA Sequencing Data using Interindividual Variation in Gene Expression. 

sessionInfo()
